src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs

```diff
@@ -1,11 +1,69 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Runtime.CompilerServices;
+
 namespace System.Reflection.Metadata
 {
     public static class AssemblyExtensions
     {
         [CLSCompliant(false)]
         public static unsafe bool TryGetRawMetadata(this Assembly assembly, out byte* blob, out int length) => throw new NotImplementedException();
+
+        /// <summary>
+        /// Updates the specified assembly using the provided metadata, IL and PDB deltas.
+        /// </summary>
+        /// <remarks>
+        /// Currently executing methods will continue to use the existing IL. New executions of modified methods will
+        /// use the new IL. Different runtimes may have different limitations on what kinds of changes are supported,
+        /// and runtimes make no guarantees as to the state of the assembly and process if the delta includes
+        /// unsupported changes.
+        /// </remarks>
+        /// <param name="assembly">The assembly to update.</param>
+        /// <param name="metadataDelta">The metadata changes to be applied.</param>
+        /// <param name="ilDelta">The IL changes to be applied.</param>
+        /// <param name="pdbDelta">The PDB changes to be applied.</param>
+        /// <exception cref="ArgumentNullException">The assembly argument is null.</exception>
+        /// <exception cref="NotSupportedException">The update could not be applied.</exception>
+        public static void ApplyUpdate(Assembly assembly, ReadOnlySpan<byte> metadataDelta, ReadOnlySpan<byte> ilDelta, ReadOnlySpan<byte> pdbDelta = default)
+        {
+#if !FEATURE_METADATA_UPDATE
+            throw new NotSupportedException ("Method body replacement not supported in this runtime");
+#else
+            if (assembly == null)
+            {
+                throw new ArgumentNullException(nameof(assembly));
+            }
+
+            RuntimeAssembly? runtimeAssembly = assembly as RuntimeAssembly;
+            if (runtimeAssembly == null)
+            {
+                throw new ArgumentException(SR.Argument_MustBeRuntimeAssembly);
+            }
+
+            unsafe
+            {
+                IntPtr monoAssembly = runtimeAssembly.GetUnderlyingNativeHandle ();
+                fixed (byte* metadataDeltaPtr = metadataDelta, ilDeltaPtr = ilDelta, pdbDeltaPtr = pdbDelta)
+                {
+                    ApplyUpdate_internal(monoAssembly, metadataDeltaPtr, metadataDelta.Length, ilDeltaPtr, ilDelta.Length, pdbDeltaPtr, pdbDelta.Length);
+                }
+            }
+#endif
+        }
+
+        internal static void ApplyUpdateSdb(Assembly assembly, byte[] metadataDelta, byte[] ilDelta, byte[]? pdbDelta = null)
```

Comment:
Why is this method needed?

Reply:
I don't know how to call a method taking a `ReadOnlySpan<byte>` argument using debugger-libs.

Reply:
Shouldn't this code be in debugger-libs when ApplyUpdate is public API?

Reply:
The issue is that we don't want to add a debugger-libs command for hot reload (we could do it, but previously there didn't seem to be a need).  So we're using the existing "create a byte[] blob" and "invoke a function" debugger-libs commands.  The "invoke a function" debugger-libs command can't use `ROS<byte>` arguments - at least I don't know how to do it.

So there are three options:
1. Call a function with better arguments (what i'm doing in this PR)
2. or, add a soft debugger protocol command for applying a hot reload delta.
3. Add a way to pass `ref struct` arguments when invoking using the debugger.

Reply:
Why cannot we do "create a byte[] blob" and "invoke a function" inside debugger-libs? Either way, this method cannot be in a permanent descriptor or the hot reload will always ship with release config

Reply:
We will probably need to add a new command to the debugger protocol, then.

Can I keep this for now?

Reply:
I'd rather not as it's gonna screw our size tracking numbers

Reply:
Discussed offline: this method is (temporarily) replacing an existing LoadMetadataUpdate private API so the size tracking isn't affected by this PR.  I added an issue to switch to debugger libs and remove ApplyUpdateSdb.  But we're going to keep it for now to unblock downstream work